Setup CI/CD workflow: linting, testing, docker build

DimaMakovskyi opened the pull request:
Description:

This pull request introduces the initial setup for the CI/CD pipeline in the Django project.
It includes the following steps:

 Linting via flake8

Testing via manage.py test

Docker image build using Dockerfile

GitHub Actions integration via .github/workflows/ci-cd.yml

This implementation ensures automated code checks, builds, and test runs on every push to main.

mehalyna commented on .github/workflows/ci-cd.yml:
```diff
@@ -0,0 +1,109 @@
+        with:
+          python-version: '3.12'
+
+      - name: Install dependencies
```
You may want to cache pip dependencies for faster builds. See GitHub's setup-python action docs for pip caching strategies.

mehalyna commented on .github/workflows/ci-cd.yml:
```diff
@@ -0,0 +1,109 @@
+
+      - name: Run Migrations
+        env:
+          DJANGO_SETTINGS_MODULE: core.settings
+          SECRET_KEY: dummy_key_for_ci
+          DEBUG: 0
+          DB_NAME: test_db
+          DB_USER: postgres
+          DB_PASSWORD: postgres
+          DB_HOST: localhost
```
You're repeating environment variables for migration and testing steps. You might consider using env at the job level to avoid duplication and make updates easier.

mehalyna commented on .github/workflows/ci-cd.yml:
```diff
@@ -0,0 +1,109 @@
+
+      - name: Run Tests
+        env:
+          DJANGO_SETTINGS_MODULE: core.settings
+          SECRET_KEY: dummy_key_for_ci
+          DEBUG: 0
+          DB_NAME: test_db
+          DB_USER: postgres
+          DB_PASSWORD: postgres
+          DB_HOST: localhost
```
You're repeating environment variables for migration and testing steps. You might consider using env at the job level to avoid duplication and make updates easier.

DimaMakovskyi replied:
fix

mehalyna commented on .github/workflows/ci-cd.yml:
```diff
@@ -0,0 +1,109 @@
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+
+      # Temporarily disabled Docker Hub login
+      # - name: Log in to Docker Hub
+      #   uses: docker/login-action@v3
+      #   with:
+      #     username: ${{ secrets.DOCKER_USERNAME }}
+      #     password: ${{ secrets.DOCKER_PASSWORD }}
```
Docker Hub login steps are commented out. Is this intentional for now? If you plan to enable these later, consider adding a TODO comment to avoid forgetting.

DimaMakovskyi replied:
fix

mehalyna commented on .github/workflows/ci-cd.yml:
```diff
@@ -0,0 +1,109 @@
+        run: |
+          docker build -t my-django-app:latest .
+
+      # Temporarily disabled push to Docker Hub
+      # - name: Push Docker Image
+      #   run: |
+      #     docker push ${{ secrets.DOCKER_USERNAME }}/projectname:latest
```
Docker image push steps are commented out. Is this intentional for now? If you plan to enable these later, consider adding a TODO comment to avoid forgetting.

DimaMakovskyi replied:
fix

mehalyna commented on .github/workflows/ci-cd.yml:
```diff
@@ -0,0 +1,109 @@
+      #     username: ${{ secrets.DOCKER_USERNAME }}
+      #     password: ${{ secrets.DOCKER_PASSWORD }}
+
+      - name: Build Docker Image
```
You might want to tag your docker image with the commit SHA or GitHub run number in addition to latest. This makes it easier to trace images back to code changes.

DimaMakovskyi replied:
fix

mehalyna commented on .github/workflows/ci-cd.yml:
```diff
@@ -0,0 +1,109 @@
+      #     docker push ${{ secrets.DOCKER_USERNAME }}/projectname:latest
+
+  deploy:
+    if: false  # Deployment is disabled during development
```
Deployment is currently disabled. Consider using an environment variable or workflow input to toggle deployment on/off for flexibility.

DimaMakovskyi commented:
This pull request addresses feedback and introduces several improvements to the CI/CD pipeline setup:

Moved duplicate environment variables to the job-level scope to reduce repetition.

Added TODO comments for Docker login and image push steps for future enablement.

Tagged Docker images with both latest and commit SHA for traceability.

 Added workflow_dispatch and an enable_deploy input to toggle deployment via manual trigger.

Used pip cache for faster dependency installation in GitHub Actions.

These changes improve the maintainability and flexibility of the CI/CD workflow.
